honeybee_energy/writer.py

```diff
@@ -22,44 +28,368 @@ def generate_idf_string(object_type, values, comments=None):
+            year_schedule, week_schedules = sched.to_idf()
+            sched_strs.extend([year_schedule] + week_schedules + day_scheds)
+        except AttributeError:  # ScheduleFixedInterval
+            if sched_dir is None:
+                sched_dir = schedule_directory if schedule_directory is not None \
+                    else os.path.join(os.environ['USERPROFILE'], 'honeybee',
+                                      'unnamed', 'schedules')
+            sched_strs.append(sched.to_idf(sched_dir))
+        t_lim = sched.schedule_type_limit
+        if t_lim is not None and not _instance_in_array(t_lim, type_limits):
+            type_limits.append(t_lim)
+    model_str.append('!-   ========= SCHEDULE TYPE LIMITS =========\n')
+    model_str.extend([type_limit.to_idf() for type_limit in set(type_limits)])
+    model_str.append('!-   ============== SCHEDULES ==============\n')
+    model_str.extend(sched_strs)
+
+    # write all of the materials and constructions
+    materials = []
+    construction_strs = []
+    for constr in model.properties.energy.constructions:
+        try:
+            materials.extend(constr.materials)
+            construction_strs.append(constr.to_idf())
+        except AttributeError:
+            pass  # ShadeConstruction; No need to write to IDF
+    model_str.append('!-   ============== MATERIALS ==============\n')
+    model_str.extend([mat.to_idf() for mat in set(materials)])
+    model_str.append('!-   ============ CONSTRUCTIONS ============\n')
+    model_str.extend(construction_strs)
+
+    # write all of the zone geometry
+    model_str.append('!-   ============ ZONE GEOMETRY ============\n')
+    for room in model.rooms:
+        model_str.append(room.to.idf(room))
+        for face in room.faces:
+            model_str.append(face.to.idf(face))
+            for ap in face.apertures:
+                model_str.append(ap.to.idf(ap))
+                for shade in ap.outdoor_shades:
+                    model_str.append(shade.to.idf(shade))
+            for dr in face.doors:
+                model_str.append(dr.to.idf(dr))
+            for shade in face.outdoor_shades:
+                model_str.append(shade.to.idf(shade))
+        for shade in room.outdoor_shades:
+            model_str.append(shade.to.idf(shade))
+
+    # write all context shade geometry
+    model_str.append('!-   ========== CONTEXT GEOMETRY ==========\n')
+    for shade in model.orphaned_shades:
+        model_str.append(shade.to.idf(shade))
+
+    return '\n\n'.join(model_str)
+
+
+def _instance_in_array(object_instance, object_array):
+    """Check if a specific object instance is already in an array.
+
+    This can be much faster than  `if object_instance in object_arrary`
+    when you expect to be testing a lot of the same instance of an object for
+    inclusion in an array since the builtin method uses an == operator to
+    test inclusion.
+    """
+    for val in object_array:
```

Comment:
If you are using it too many times than it might sense to create a `set` from `object_instance` and do all your inclusion checks.

Reply:
I do eventually create a set (a few lines after this) but there's a very good performance reason why I use this method first.

Let's say I have a model of 1000 Faces and each one has a construciton that needs to be written into EnergyPlus. 500 of these Faces actually have the same "interior wall" construction instance assigned to them.  If I make a set of all 1000 construciton instances, it's going to check whether each of those constructions is equal to any others in the set. This is really computationally intensive given that each and every property of each and every material in the construction is checked for equality. If I simply check whether the object instance is already in the array, it is orders of magnitude faster (the performance tests I ran showed that it ran 1000 times faster). So that is why we use this method here.